### ISSUE TITLE
postgres 10 - table partitioning

### ISSUE THREAD

**Issue by contra** [sequelize/sequelize#7673]
https://wiki.postgresql.org/wiki/New_in_postgres_10#Native_Partitioning

Two questions:

- Currently, is there any way to modify/configure the command executed when creating tables to support adding partitioning to a table?
- In the future, will there be an option on a table to support partitioning?

Open to sending a PR but this looks like a big feature, so not sure I'd be able to do it as fast as you.

**Comment by contra** [comment]
Yep, keep it open

**Comment by contra** [comment]
Keep it open

**Comment by dskrvk** [comment]
Also interested in this.

**Comment by gijo-varghese** [comment]
@contra +1

Did anyone find a solution for this?

**Comment by contra** [comment]
I'm going to try a PR that will allow you to specify partitioning and partition type on table create, but partition management will still be a manual process since they don't get created automatically anyways.

**Comment by sushantdhiman** [comment]
Ah, I wasn't aware of this issue, I have added a feature flag.

Looks like MSSQL, MySQL and Postgres all supports partitioning. Its going to be somewhat large work because we need at-least two features

1. defining partitions on tables
2. referring them in queries

An experimental implementation can be found [here](https://github.com/sequelize/sequelize/pull/8995) I would like to implement support for all three dialect when we do implement this feature.

This feature needs some API discussion that covers all three dialects. I have not used partitioning so any ideas are appreciated.

This feature will likely to be implemented after v5 is released

**Comment by dskrvk** [comment]
Our team would definitely use this.

**Comment by papb** [comment]
@contra What is the status on this?

**Comment by contra** [comment]
I sent a PR (https://github.com/sequelize/sequelize/pull/8995) that was declined, so this feature is not being implemented by me. I believe @sushantdhiman is implementing this.

**Comment by hlongvu** [comment]
Also interested in this. Will this feature be added?

**Comment by myrewin** [comment]
I need this badly now, any hope ?

**Comment by yeongjet** [comment]
Use [hooks](https://sequelize.org/v5/manual/hooks.html) maybe helpful
```ts
@Table({
  tableName: 'person'
})
export class Person extends Model<Person> {

  @PrimaryKey
  @AutoIncrement
  @Column(DataType.INTEGER)
  id: number;

  @AllowNull(false)
  @Column(DataType.JSONB)
  name: string

  @AllowNull(false)
  @Column(DataType.JSONB)
  age: number

  @BeforeCreate({name: 'beforeCreateHook'})
  @BeforeBulkCreate({name: 'beforeBulkCreateHook'})
  @BeforeUpdate({name: 'beforeUpdateHook'})
  static beforeCreateHooks(instance: Person, options: any) {
    // caculate which table use
   const x = 1
    _.set(instance.sequelize.models.Person, 'tableName', `person_${x}`);
  }
  
  @BeforeFind({name: 'beforeFindHook'})
  static beforeFindHook(options: any) {
    _.set(this, 'tableName', `${this.tableName}_${options.partion}`)
  }
}
```
usage:
```ts
 await Person.findOne({
    where: {
        age: 2
    },
    partion: 2
 })
```

**Comment by ahuangfeng** [comment]
Interested

**Comment by SeanSanker** [comment]
very interested

**Comment by changmyeong** [comment]
> Use [hooks](https://sequelize.org/v5/manual/hooks.html) maybe helpful
> 
> ```ts
> @Table({
>   tableName: 'person'
> })
> export class Person extends Model<Person> {
> 
>   @PrimaryKey
>   @AutoIncrement
>   @Column(DataType.INTEGER)
>   id: number;
> 
>   @AllowNull(false)
>   @Column(DataType.JSONB)
>   name: string
> 
>   @AllowNull(false)
>   @Column(DataType.JSONB)
>   age: number
> 
>   @BeforeCreate({name: 'beforeCreateHook'})
>   @BeforeBulkCreate({name: 'beforeBulkCreateHook'})
>   @BeforeUpdate({name: 'beforeUpdateHook'})
>   static beforeCreateHooks(instance: Person, options: any) {
>     // caculate which table use
>    const x = 1
>     _.set(instance.sequelize.models.Person, 'tableName', `person_${x}`);
>   }
>   
>   @BeforeFind({name: 'beforeFindHook'})
>   static beforeFindHook(options: any) {
>     _.set(this, 'tableName', `${this.tableName}_${options.partion}`)
>   }
> }
> ```
> 
> usage:
> 
> ```ts
>  await Person.findOne({
>     where: {
>         age: 2
>     },
>     partion: 2
>  })
> ```

it's a good solution.
but it will change table name multiple times. and when context change to concurrency, it is impossible to know which partition to use.

**Comment by pnutmath** [comment]
Are we moving ahead with this? I am interested in it

**Comment by williamfeng323** [comment]
is there anyone working on this?

**Comment by contra** [comment]
I submitted a PR but the maintainers said there wasn't enough demand for it and closed it - https://github.com/sequelize/sequelize/pull/8995

**Comment by WikiRik** [comment]
Hi all! I think that we can accept PRs for this but it's not a priority for the current maintainers to implement this ourselves.
So feel free to make a PR for this. Preferably this would not only be for postgres, but also mysql and mssql as per https://github.com/sequelize/sequelize/issues/7673#issuecomment-364335413

**Comment by yocontra** [comment]
@WikiRik Is the syntax I proposed in my PR acceptable? If so, it can be reopened or redone on latest and support for other drivers can be added.

**Comment by yohaansunnie** [comment]
Is anyone looking into @yocontra PR?
Really need this feature.

**Comment by ephys** [comment]
Looked into the PR. It's a good start and thank you for doing the initial work.

I would rename the `partition` option to `partitionBy`

As for the type it accepts ; if we're going to add it, we need to make sure it's compatible with the other dialects that support partitioning. Below is my initial research.

## postgres

The full [postgres](https://www.postgresql.org/docs/14/ddl-partitioning.html) Syntax for `PARTITION BY` is

```
[ PARTITION BY { RANGE | LIST | HASH } ( { column_name | ( expression ) } [ COLLATE collation ] [ opclass ] [, ... ] ) ]
```

## mysql

[MySQL](https://dev.mysql.com/doc/refman/8.0/en/partitioning-overview.html) seems to work similarly to postgres, with more options

```
partition_options:
    PARTITION BY
        { [LINEAR] HASH(expr)
        | [LINEAR] KEY [ALGORITHM={1 | 2}] (column_list)
        | RANGE{(expr) | COLUMNS(column_list)}
        | LIST{(expr) | COLUMNS(column_list)} }
    [PARTITIONS num]
    [SUBPARTITION BY
        { [LINEAR] HASH(expr)
        | [LINEAR] KEY [ALGORITHM={1 | 2}] (column_list) }
      [SUBPARTITIONS num]
    ]
    [(partition_definition [, partition_definition] ...)]
```

## mssql

I'm not sure I understand how [mssql partitions](https://docs.microsoft.com/en-us/sql/relational-databases/partitions/create-partitioned-tables-and-indexes?view=sql-server-ver15) work.

If someone can explain the syntax we'd need to support, it would greatly help me.

## mariadb

Documentation: https://mariadb.com/kb/en/partitioning-overview/

## snowflake

Not sure if it's supported, couldn't find much information

## sqlite

Probably not supported, couldn't find much information

## db2

https://www.ibm.com/docs/en/db2/11.5?topic=tables-table-partitioning

Looks similar to postgres / mariadb / mysql

## ibmi

https://www.ibm.com/docs/en/i/7.4?topic=tables-creation-partitioned

Looks similar to postgres / mariadb / mysql

# Initial Conclusion

Given the differences between the different dialects, I think it may be easier to handle this using raw SQL:

```typescript
import { Model, literal } from '@sequelize/core';

class User extends Model {}

User.init({ /* attributes */ }, {
  sequelize,
  partitionBy: literal('LIST ("sourceId")'),
})
```

**Comment by lethak** [comment]
This kind of feature is mandatory for any serious database heavy project. I would love to see it managed by sequelize

**Comment by kamaladenalhomsi** [comment]
As @lethak mentioned, this is very very mandatory.

**Comment by zia-mazari** [comment]
I have a table of 13.6+ Million rows, in the table, there are 2 columns with JSONB(array of objects) with more or less 10 key-value pairs. 
Query filters are applied to almost every column, even on the JSONB columns.

I'm using Sequelize ORM and it's working with 6+ Million rows but not for 13.6+ Million. I also implemented indexes properly but still, there are some queries that are taking time. 

Now I'm working on table partitioning manually. Sequelize should add this feature.

**Comment by rahulchander2Y23** [comment]
Yep - very interested in this feature!  Any ETA on the development?

**Comment by hammadassistalong** [comment]
Very interested in this feature, its been almost 5 years. Do we have to die and re-live for this feature?

**Comment by nullgrim** [comment]
Interested in this feature, but I am unsure how it isn't a thing yet.. been so long.

**Comment by cskiwi** [comment]
Would love to have this build in to sequelize. will try the hook method anyway :)

**Comment by mihairadulescu** [comment]
Any updates on this one ?

**Comment by edutedu** [comment]
Very interested in this feature

**Comment by Unodetantosluises** [comment]
Any update on this feature, im interested

**Comment by nahidiftekhar** [comment]
kinda essential for large scale projects. Would love to see it added.

**Comment by MatviiStelmakh** [comment]
Any progress on this feature? We currently need it.

**Comment by hkaranjule77** [comment]
Any update on this feature?

**Comment by devalex365** [comment]
I'm interested in this feature as well, still no updates?

**Comment by MichaelTendoSsemwanga** [comment]
+1
